### PR TITLE
Various Fixes and Updates

### LINKED ISSUE #1722: Custom Factor in MATLAB
I am currently engaged in an effort to integrate an existing factor in GTSAM, namely the _BarometricFactor_, into the latest MATLAB wrapper. To this end, I have had to modify both the _BarometricFactor.cpp_ and _BarometricFactor.h_ files, ensuring their proper compilation.

However, I've encountered a stumbling block while attempting to wrap these modifications for MATLAB integration. A parsing error arises during this phase, reminiscent of a similar issue I found documented online (Issue #1673 regarding error compiling a new class and GTSAM_EXPORT to MATLAB), which was attributed to minor syntax errors within the _example.i_ interface file.

For your convenience, I've attached the modified _.cpp_ and _.h_ files, alongside the _.i_ file and a screenshot capturing the specific error encountered in the terminal. Any insight or expertise into resolving this matter would be invaluable and greatly appreciated.

image

**example.i**
 
```
// Specify classes from gtsam that we are using
virtual class gtsam::Value;
class gtsam::Point3;
class gtsam::Pose3;
virtual class gtsam::noiseModel::Base;
virtual class gtsam::NonlinearFactor;
virtual class gtsam::NoiseModelFactor2;
class gtsam::NonlinearFactorGraph;
class gtsam::Values;
class gtsam::Key;

/// Include custom factor
#include <example/BarometricFactor.h>

namespace example {

virtual class BarometricFactor : gtsam::NonlinearFactor {
  
  /// Standard Constructor
  BarometricFactor();
  BarometricFactor(
    size_t key, 
    size_t baroKey, 
    double double& baroIn, 
    const gtsam::noiseModel::Base* model);

  /// Testable
  void print(string s = "", const gtsam::KeyFormatter& keyFormatter = gtsam::DefaultKeyFormatter) const;
  bool equals(const gtsam::NonlinearFactor& expected, double tol);

  /// Standard Interface
  const double& measurementIn() const;
  double heightOut(double n) const;
  double baroOut(const double& meters) const;
};

} /// namespace example
```

**BarometricFactor.h**
 
```
#pragma once

#include <iostream>
#include <memory>
#include <string>
#include <boost/optional.hpp>
#include <gtsam/nonlinear/NonlinearFactor.h>
#include <gtsam/geometry/Pose3.h>

using namespace gtsam;

namespace example {

class BarometricFactor : public NoiseModelFactorN<Pose3,double> {

private:

    /// Basis to form custom unary factor
    typedef NoiseModelFactorN<Pose3,double> Base;

    /// Input
    double nT_;

public:

    /// Provide access to the Matrix& version of evaluateError
    using Base::evaluateError;

    /// shorthand for a smart pointer to a factor
    typedef std::shared_ptr<BarometricFactor> shared_ptr;

    /// Typedef to this class
    typedef BarometricFactor This;

    /** default constructor - only use for serialization */
    BarometricFactor(): nT_(0) {}
    ~BarometricFactor() override {}

    /// Barometric Update
    BarometricFactor(Key key, Key baroKey, const double& baroIn, const SharedNoiseModel& model) :
        Base(model,key,baroKey),nT_(heightOut(baroIn)) {}

    /// @return a deep copy of this factor
    gtsam::NonlinearFactor::shared_ptr clone() const override {
        return std::static_pointer_cast<gtsam::NonlinearFactor>(
            std::shared_ptr<This>(new This(*this)));}
    
    /// print
    void print(
        const std::string& s = "", 
        const KeyFormatter& keyFormatter = DefaultKeyFormatter) const override;

    /// equals
    bool equals(const NonlinearFactor& expected, double tol = 1e-9) const override;

    /// error evaluation
    Vector evaluateError(const Pose3& p, const double& bias, 
             OptionalMatrixType H, OptionalMatrixType H2) const override; 

    /// Input
    inline const double& measurementIn() const { return nT_; }
    
    /// Earth Atmosphere Model
    inline double heightOut(double n) const {
        return (std::pow(n / 101.29, 1. / 5.256) * 288.08 - 273.1 - 15.04) / -0.00649;}
    
    /// Pressure Estimation
    inline double baroOut(const double& meters) const{
        double temp = 15.04 - 0.00649 * meters;
        return 101.29 * std::pow(((temp + 273.1) / 288.08), 5.256);}

};
} // namespace example

```

**BarometricFactor.cpp**

```
#include "BarometricFactor.h"
#include <gtsam/inference/Key.h>
#include <gtsam/linear/NoiseModel.h>
#include <gtsam/base/Vector.h> 
#include <iostream>

using namespace std;

namespace example{

 //***************************************************************************
 void BarometricFactor::print(const string& s,
                              const KeyFormatter& keyFormatter) const {
     cout << (s.empty() ? "" : s + " ") << "Barometric Factor on "
          << keyFormatter(key<1>()) << "Barometric Bias on "
          << keyFormatter(key<2>()) << "\n";
  
     cout << "  Baro measurement: " << nT_ << "\n";
        noiseModel_->print("  noise model: ");
 }
  
 //***************************************************************************
 bool BarometricFactor::equals(const NonlinearFactor& expected,
                               double tol) const {
        const This* e = dynamic_cast<const This*>(&expected);
        return e != nullptr && Base::equals(*e, tol) &&
            traits<double>::Equals(nT_, e->nT_, tol);
 }
  
 //***************************************************************************
 Vector BarometricFactor::evaluateError(
                                        const Pose3& p, 
                                        const double& bias,
                                        OptionalMatrixType H,
                                        OptionalMatrixType H2) const {
        Matrix tH;
        Vector ret = (Vector(1) << (p.translation(tH).z() + bias - nT_)).finished();
        if (H) (*H) = tH.block<1, 6>(2, 0);
        if (H2) (*H2) = (Matrix(1, 1) << 1.0).finished();
        return ret;
 }
 };
```

### PR DESCRIPTION
**Pull request by varunagrawal**
1. Undo Python CI changes since the intended effects are not being purported. The Python-TBB CI needs more investigation to debug what the underlying issue is rather than making random guesses.
2. Fix boost collision issue for `save`, `load`, etc with `boost/serialization/optional.hpp`.
3. Wrapped the `BarometricFactor` as required in #1722.

I ran this locally on my machine Matlab and it works great!

### CODE REVIEW

**Comment by talregev** [comment]
It not fair that you cancel my PR, but you use my changes. Be fair.

**Comment by varunagrawal** [comment]
@talregev your PR didn't clearly state what the issue you were trying to solve was. I encountered the issue myself this week and recalled your fix.
If you want your PRs to be accepted and merged, you need to better communicate what the issue is and how you're solving it. As a reviewer, I have limited time and the onus is on you to [write a good PR](https://developers.google.com/blockly/guides/contribute/get-started/write_a_good_pr).

**Comment by talregev** [comment]
Even you think I wasn't communicate enough, and then you recall my fix, Do you think is a reason to steal it like that? 

I will put my description here of my PR that the community will judge if it was communicate enough.

> Bug fix for macos: Support new version of boost along side old version. #1721
In macos vm, recently we get a newer version of boost (1.84).
They already implemented in the new version the function that gtsam implemented.
So I am not sure the correct version of boost that it begin to happen, so I put version 1.80.
We can tune the version later (when we have another issue similar to that in different boost version).
@varunagrawal @ProfFan
Thank you for your time and effort for review this PR.

**Comment by varunagrawal** [comment]
Sounds good.

I'm re-reading your comment and it clearly doesn't mention what issue your PR is trying to fix. There is some vague mention of the MacOS VM but nothing about the fact that the latest boost version is causing a name collision with std::optional.

I looked up the error and the suggested fix was similar to yours which is why I recalled your PR. There was no ill intention and I didn't have the time to reopen your PR and go through the review back and forth, hence why I added it to this PR.

**Comment by varunagrawal** [comment]
The fix on macOS doesn't seem to generalize to Ubuntu. I'll look into it later today.

**Comment by varunagrawal** [comment]
@ProfFan FYI this PR has a better analysis of the issue that #1721 is trying to fix and thus has the better solution.

File: .github/workflows/build-python.yml [hunk 1]
```diff
@@ -29,7 +29,7 @@ jobs:
         name:
           [
             ubuntu-20.04-gcc-9,
-            ubuntu-22.04-gcc-9-tbb,
```
Comment (ProfFan) [inline .github/workflows/build-python.yml:32]: Why this change?

**Review by ProfFan — APPROVED** [review]
Sans one comment

File: .github/workflows/build-python.yml (hunk 1 above)
Comment (varunagrawal) [reply .github/workflows/build-python.yml:32]: This change was made in a previous PR to fix the CI hanging on Python+TBB, however it doesn't seem to have made a difference as the issue still persists.

A lot of people still use Ubuntu 20.04 since it is the last version that supports ROS1 so having a CI environment for it makes more sense (plus backwards compatibility).
